Adds ability to take an ABI to handle `EntryFunction` JSON arguments correctly

gupnik opened the pull request:
## Description

Fixes https://github.com/trustwallet/wallet-core/issues/3899

This PR adds the ability to accept an ABI in the `SigningInput` so that we are able to parse the arguments correctly.

## How to test

Run Rust Tests

## Types of changes

Add `abi` in `tw_proto::Aptos::Proto::SigningInput`.

## Checklist

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

satoshiotomakan commented on src/proto/Aptos.proto:
```diff
@@ -166,6 +166,8 @@ message SigningInput {
     TokenTransferCoinsMessage token_transfer_coins = 15;
   }
+
+  optional string abi = 16;
```
You can keep 5 indexes for the `transaction_payload` new variants that can come in the future.
So `abi` can be 21st field

satoshiotomakan reviewed (changes requested):
@gupnik thanks for the PR! Please consider the following notes

gupnik commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -99,6 +115,41 @@ impl TryFrom<Value> for EntryFunction {
+    } else if let Value::Array(list) = val {
+        let vals = list
+            .into_iter()
+            .map(|v| serde_json::from_value::<u8>(v).map_err(|_| anyhow::anyhow!("expected u8")))
```
> We need to handle other array inner types, not `u8` only https://github.com/aptos-labs/aptos-core/blob/0e4f5df59f7518ddd82b9116a7cb942bfa658741/api/types/src/convert.rs#L915

```rust
#[derive(Clone, Hash, Eq, PartialEq, Serialize, Deserialize)]
pub enum TransactionArgument {
    U8(u8),
    U64(u64),
    U128(u128),
    Address(AccountAddress),
    U8Vector(#[serde(with = "serde_bytes")] Vec<u8>),
    Bool(bool),
    // NOTE: Added in bytecode version v6, do not reorder!
    U16(u16),
    U32(u32),
    U256(u256::U256),
}
```

`TransactionArgument` from `move-core-types` can actually handle only `U8Vector`. So, others should not occur in our context?

satoshiotomakan replied:
Makes sense, thanks

satoshiotomakan replied:
Needs to be fixed to be able to handle `vector<vector<u8>>` and other complex argument types

satoshiotomakan replied:
So you may need to add our own `TransactionArgument` type, or reuse `MoveType`

10gic commented:
Hi @gupnik ,

Thanks for the PR! Aptos has its own ABI definition. For example: [Abi example](https://explorer.aptoslabs.com/account/0xc23c3b70956ce8d88fb18ad9ed3b463fe873cb045db3f6d2e2fb15b9aab71d50/modules/code/IFO?network=mainnet).

Can we use the office ABI format instead? I'm concerned that the simple ABI introduced in this PR may not cover more complex cases.

gupnik commented:
> Hi @gupnik ,
> 
> Thanks for the PR! Aptos has its own ABI definition. For example: [Abi example](https://explorer.aptoslabs.com/account/0xc23c3b70956ce8d88fb18ad9ed3b463fe873cb045db3f6d2e2fb15b9aab71d50/modules/code/IFO?network=mainnet).
> 
> Can we use the office ABI format instead? I'm concerned that the simple ABI introduced in this PR may not cover more complex cases.

Hi @10gic, thanks for the comment.

As the user needs to provide an individual payload anyways to blind sign, I am not sure if we need the complete ABI parser. The current PR just expects the types of the arguments and ensures that they are parsed to `TransactionArgument`s correctly.

So, the caller should simply be able to retrieve the ABI via the API: https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/{address}/modules and call the function with the respective types.

10gic commented:
> > Hi @gupnik ,
> > Thanks for the PR! Aptos has its own ABI definition. For example: [Abi example](https://explorer.aptoslabs.com/account/0xc23c3b70956ce8d88fb18ad9ed3b463fe873cb045db3f6d2e2fb15b9aab71d50/modules/code/IFO?network=mainnet).
> > Can we use the office ABI format instead? I'm concerned that the simple ABI introduced in this PR may not cover more complex cases.
> 
> Hi @10gic, thanks for the comment.
> 
> As the user needs to provide an individual payload anyways to blind sign, I am not sure if we need the complete ABI parser. The current PR just expects the types of the arguments and ensures that they are parsed to `TransactionArgument`s correctly.
> 
> So, the caller should simply be able to retrieve the ABI json via the API: [https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/{address}/modules](https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/%7Baddress%7D/modules) and call the function with the respective types.

We don't need to pass the entire ABI JSON file; it's sufficient to only pass the `params` field.

The `params` field can have complex cases. For example:
```
            "params": [
                "vector<vector<u8>>",
                "vector<vector<u8>>",
                "vector<u64>",
                "0x1::coin::Coin<0x1::aptos_coin::AptosCoin>"
            ]

            "params": [
                "&signer",
                "vector<vector<u8>>",
                "vector<vector<u8>>",
                "vector<u64>"
            ]

            "params": [
                "0x7e783b349d3e89cf5931af376ebeadbfab855b3fa239b7ada8f5a92fbea6b387::price_identifier::PriceIdentifier",
                "u64"
            ],
```
These examples are taken from: https://aptoscan.com/module/0x7e783b349d3e89cf5931af376ebeadbfab855b3fa239b7ada8f5a92fbea6b387/pyth

Can you add more unit tests to improve coverage?

gupnik commented:
> > > Hi @gupnik ,
> > > Thanks for the PR! Aptos has its own ABI definition. For example: [Abi example](https://explorer.aptoslabs.com/account/0xc23c3b70956ce8d88fb18ad9ed3b463fe873cb045db3f6d2e2fb15b9aab71d50/modules/code/IFO?network=mainnet).
> > > Can we use the office ABI format instead? I'm concerned that the simple ABI introduced in this PR may not cover more complex cases.
> > 
> > 
> > Hi @10gic, thanks for the comment.
> > As the user needs to provide an individual payload anyways to blind sign, I am not sure if we need the complete ABI parser. The current PR just expects the types of the arguments and ensures that they are parsed to `TransactionArgument`s correctly.
> > So, the caller should simply be able to retrieve the ABI json via the API: [https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/{address}/modules](https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/%7Baddress%7D/modules) and call the function with the respective types.
> 
> We don't need to pass the entire ABI JSON file; it's sufficient to only pass the `params` field.
> 
> The `params` field can have complex cases. For example:
> 
> ```
>             "params": [
>                 "vector<vector<u8>>",
>                 "vector<vector<u8>>",
>                 "vector<u64>",
>                 "0x1::coin::Coin<0x1::aptos_coin::AptosCoin>"
>             ]
> 
>             "params": [
>                 "&signer",
>                 "vector<vector<u8>>",
>                 "vector<vector<u8>>",
>                 "vector<u64>"
>             ]
> 
>             "params": [
>                 "0x7e783b349d3e89cf5931af376ebeadbfab855b3fa239b7ada8f5a92fbea6b387::price_identifier::PriceIdentifier",
>                 "u64"
>             ],
> ```
> 
> These examples are taken from: https://aptoscan.com/module/0x7e783b349d3e89cf5931af376ebeadbfab855b3fa239b7ada8f5a92fbea6b387/pyth
> 
> Can you add more unit tests to improve coverage?

Thanks @10gic for raising this. I went through `aptos-core` and it seems that the encoding can be implemented via the following steps:
1. Accept the abi and decode each parameter into a [MoveType](https://github.com/aptos-labs/aptos-core/blob/0e4f5df59f7518ddd82b9116a7cb942bfa658741/api/types/src/move_types.rs#L490)
2. Obtain the `TypeTag` as shows in tests [here](https://github.com/aptos-labs/aptos-core/blob/0e4f5df59f7518ddd82b9116a7cb942bfa658741/api/src/tests/converter_test.rs#L66)
3. Convert into a `MoveTypeLayout` as shown [here]https://github.com/aptos-labs/aptos-core/blob/81a2f4268b9e0f25cb6e1ce5c28bba0a34c27604/api/types/src/convert.rs#L854)
   - This internally goes to `TypeBuilder` [here](https://github.com/diem/move/blob/725168d7522a3abeeb8664b4f1498552b3657286/language/tools/move-bytecode-utils/src/layout.rs#L231) 
   - Accordingly, [this](https://github.com/diem/move/blob/725168d7522a3abeeb8664b4f1498552b3657286/language/tools/move-bytecode-utils/src/layout.rs#L261) states that an input cannot be a `signer` while others are converted appropriately
 4.  Create a `MoveValue` as shown [here](https://github.com/aptos-labs/aptos-core/blob/81a2f4268b9e0f25cb6e1ce5c28bba0a34c27604/api/types/src/convert.rs#L874) which can be appropriately serialized.

I will try to implement a similar flow and the respective unit tests for serialisation but please let me know if you would want me try to broadcast any specific transactions.

CC: @satoshiotomakan

satoshiotomakan commented:
@gupnik looks good. From my point of view, unit tests with only value parsing would be enough, if we borrow them from `aptos-core` repository. Can you please leave comments with the links to the original unit tests when finish?

gupnik commented:
> @gupnik looks good. From my point of view, unit tests with only value parsing would be enough, if we borrow them from `aptos-core` repository. Can you please leave comments with the links to the original unit tests when finish?

@satoshiotomakan I have added the implementation as discussed. Also, [here](https://github.com/aptos-labs/aptos-core/blob/0e4f5df59f7518ddd82b9116a7cb942bfa658741/api/src/tests/converter_test.rs#L17) are the tests that I have replicated as well.

To note: `aptos-core` uses a module resolver internally to obtain the struct types. In order to be able to accept that as an API, I have added the support for generics. 

For example, in the following case, `aptos-core` is able to retrieve the types of `0x1::guid::ID` and accordingly parse the input.

```rust
assert_value_conversion(
        &converter,
        "0x1::guid::ID",
        json!({"addr": "0x1", "creation_num": "1"}),
        VmMoveValue::Struct(MoveStruct::Runtime(vec![
            VmMoveValue::U64(1),
            VmMoveValue::Address(address),
        ])),
    );
```

In our case, we instead are able to accept it as follows:

```rust
assert_value_conversion(
      "0x1::guid::ID<u64, address>",
      ["1", "0x1"],
      MoveValue::Struct(MoveStruct::Runtime(vec![
          MoveValue::U64(1),
          MoveValue::Address(address),
      ])),
  );
```

satoshiotomakan commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -99,6 +123,210 @@ impl TryFrom<Value> for EntryFunction {
+        .trim_end_matches('"')
+        .to_string();
+    Ok(match layout {
+        MoveTypeLayout::Bool => MoveValue::Bool(
+            val_str
+                .parse::<bool>()
+                .map_err(|_| EncodingError::InvalidInput)?,
+        ),
```
As I can see, in `aptos-core` doesn't convert `val` into string, but parses the JSON value directly to the target type (bool in this case): https://github.com/aptos-labs/aptos-core/blob/aaa3514c8ee4e5d38b89d916eadff7286a42e040/api/types/src/convert.rs#L881

Why do we need to convert the value into a string here?

gupnik replied:
> As I can see, in `aptos-core` doesn't convert `val` into string, but parses the JSON value directly to the target type (bool in this case): https://github.com/aptos-labs/aptos-core/blob/aaa3514c8ee4e5d38b89d916eadff7286a42e040/api/types/src/convert.rs#L881
> 
> Why do we need to convert the value into a string here?

Yes, because we are passing the params as part of an outer string,
```rust
r#"[
    "vector<u8>",
    "u64",
    "bool"
 ]"#
```
we need to ensure that we trim the quotes appropriately. Hence, we are parsing the underlying string (after trimming the quotes) instead of trying to parse the Value directly (which would fail because of the presence of additional quotes).

gupnik commented:
Hi @satoshiotomakan, thanks for the review. I will go through each of them in my day tomorrow, but to clarify each line/conversion has only been added because it is needed. I will try to share how & where its used tomorrow.

gupnik commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -99,6 +123,210 @@ impl TryFrom<Value> for EntryFunction {
+    }
+}
+
+fn is_utf8_string(st: &StructTag) -> bool {
+    st.address == CORE_CODE_ADDRESS
+        && st.name.to_string() == "String"
+        && st.module.to_string() == "string"
+}
```
Used here
```rust
 MoveStructLayout::WithTypes { type_, .. } => {
            if is_utf8_string(type_) {
```

gupnik commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -273,4 +501,141 @@ mod tests {
+        );
+
+        assert_value_conversion(
+            "0x1::guid::ID<u64, address>", // As we do not have access to the module resolver, the types of the struct should be provided as params
```
> Does [https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/{address}/modules](https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/%7Baddress%7D/modules) RPC return this type argument with the type arguments like `0x1::guid::ID<u64, address>`, or it returns `0x1::guid::ID`? If it doesn't have type arguments, then the WalletCore users won't be able to provide the full ABI required by WalletCore

They should be able to use `https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/accounts/GET/accounts/{address}/resource/{resource_type}` to retrieve the types

gupnik commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -99,6 +123,210 @@ impl TryFrom<Value> for EntryFunction {
+        },
+        _ => return Err(EncodingError::InvalidInput),
+    };
+    let val = trim_if_needed(val)?;
```
> As I can see, `trim_if_needed` makes sense to use only if `val` is expected to be a string, but here below we require it to be an `Array`. So there is no need to use this function.
> 
> Also you can simply add a function like `trim_quotations(s: &str) -> &str` that takes a string, not a JSON object, and use it when needed

This is needed when array is passed to us as a part of an outer string: `"[1, 2, 3]"`

gupnik commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -99,6 +123,210 @@ impl TryFrom<Value> for EntryFunction {
+
+fn parse_struct_argument(layout: &MoveStructLayout, val: Value) -> EncodingResult<MoveValue> {
+    let field_layouts = match layout {
+        MoveStructLayout::Runtime(fields) => fields,
```
> As I can see, `aptos-core` returns an error on `MoveStructLayout::Runtime(fields)`. Could you please check if we need to handle it here?

Yes, because `aptos-core` only handles structs with named fields (that it obtains from the module resolver) whereas we are processing the unnamed fields that are provided in the API.

satoshiotomakan replied:
Makes sense, thanks!

gupnik commented on rust/chains/tw_aptos/src/transaction_payload.rs:
```diff
@@ -99,6 +123,210 @@ impl TryFrom<Value> for EntryFunction {
+            if is_utf8_string(type_) {
+                let string = val.as_str().ok_or(EncodingError::InvalidInput)?;
+                return Ok(new_vm_utf8_string(string));
+            } else {
+                return Err(EncodingError::InvalidInput);
+            }
```
> `aptos-core` allows the struct tag to be a non-utf8 string, but any other.

Yeah, `aptos-core` handles strings and structs with named fields, whereas we handle strings and structs with unnamed fields due to the reasoning mentioned here.
